// @azure/search-documents/dist-esm/src/generated/service/searchServiceClient.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import * as coreHttp from "@azure/core-http";
import { DataSources, Indexers, Skillsets, SynonymMaps, Indexes } from "./operations";
import * as Parameters from "./models/parameters";
import * as Mappers from "./models/mappers";
import { SearchServiceClientContext } from "./searchServiceClientContext";
/** @internal */
export class SearchServiceClient extends SearchServiceClientContext {
    /**
     * Initializes a new instance of the SearchServiceClient class.
     * @param endpoint The endpoint URL of the search service.
     * @param apiVersion Api Version
     * @param options The parameter options
     */
    constructor(endpoint, apiVersion, options) {
        super(endpoint, apiVersion, options);
        this.dataSources = new DataSources(this);
        this.indexers = new Indexers(this);
        this.skillsets = new Skillsets(this);
        this.synonymMaps = new SynonymMaps(this);
        this.indexes = new Indexes(this);
    }
    /**
     * Gets service level statistics for a search service.
     * @param options The options parameters.
     */
    getServiceStatistics(options) {
        const operationArguments = {
            options: coreHttp.operationOptionsToRequestOptionsBase(options || {})
        };
        return this.sendOperationRequest(operationArguments, getServiceStatisticsOperationSpec);
    }
}
// Operation Specifications
const serializer = new coreHttp.Serializer(Mappers, /* isXml */ false);
const getServiceStatisticsOperationSpec = {
    path: "/servicestats",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ServiceStatistics
        },
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [Parameters.endpoint],
    headerParameters: [Parameters.accept, Parameters.xMsClientRequestId],
    serializer
};
